### DISCUSSION TITLE
How is Kconfig output used by CMake?

### DISCUSSION THREAD

**Discussion by HumblePresent** [zephyrproject-rtos/zephyr#47510] [Q&A] [answered]
These are more of a question on the implementation details of the Zephyr build system than how to use it. As I understand from the documentation, the `.config` file generated from parsing the various Kconfig files is used by CMake to access the `CONFIG_...` definitions. As far as I can tell from the `CMakeLists.txt` file in the zephyr root directory those configuration definitions are already available by the time that file is processed. How and when is the `.config` file, which I believe is in Makefile format, translated and included in the CMake build process?

My second question is on the `autoconf.h` header also generated from Kconfig. The documentation says that the definitions in that header are automatically exposed at compile time so it doesn't need to be included in source files. How are those definitions exposed without including the header in any source files?

**Comment by yashi** [comment, marked as answer]
> **How and when is the .config file, which I believe is in Makefile format, translated and included in the CMake build process?**

It's here:
https://github.com/zephyrproject-rtos/zephyr/blob/zephyr-v3.1.0/cmake/modules/kconfig.cmake#L325-L326

> **How are those definitions exposed without including the header in any source files?**

Compilers have a flag for special include files.  [For GCC](https://gcc.gnu.org/onlinedocs/gcc/Preprocessor-Options.html),

> `-imacros file`
> Exactly like -include, except that any output produced by scanning file is thrown away. Macros it defines remain defined. This allows you to acquire all the macros from a header without also processing its declarations.
> 
> All files specified by -imacros are processed before all files specified by -include.

https://github.com/zephyrproject-rtos/zephyr/blob/zephyr-v3.1.0/CMakeLists.txt#L253-L254

**Reply by yashi** [reply]
BTW, `.config` isn't Makefile format. it's just key-value pairs.
